backend/de.metas.materialtracking/src/main/java/de/metas/materialtracking/impl/MaterialTrackingInvoiceCandService.java

```diff
@@ -0,0 +1,121 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.materialtracking.impl;
+
+import de.metas.invoicecandidate.InvoiceCandidateHeaderAggregationId;
+import de.metas.invoicecandidate.api.IAggregationBL;
+import de.metas.materialtracking.IMaterialTrackingBL;
+import de.metas.materialtracking.IMaterialTrackingDAO;
+import de.metas.materialtracking.MTLinkRequest;
+import de.metas.materialtracking.MaterialTrackingId;
+import de.metas.materialtracking.model.I_C_Invoice_Candidate;
+import de.metas.materialtracking.model.I_M_Material_Tracking;
+import de.metas.util.Services;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+@Service
+@RequiredArgsConstructor
+public class MaterialTrackingInvoiceCandService
+{
+	private final MaterialTrackingInvoiceCandRepository repo;
+	private final IAggregationBL aggregationBL = Services.get(IAggregationBL.class);
+	private final IMaterialTrackingBL materialTrackingBL = Services.get(IMaterialTrackingBL.class);
+	private final IMaterialTrackingDAO materialTrackingDAO = Services.get(IMaterialTrackingDAO.class);
+
+	public void unlinkMaterialTrackings(@NonNull final List<I_C_Invoice_Candidate> candidate, final boolean isClearAggregationKeyOverride)
+	{
+		candidate.forEach(c -> unlinkMaterialTracking(c, isClearAggregationKeyOverride));
+	}
+
+	public void unlinkMaterialTracking(@NonNull final I_C_Invoice_Candidate candidate, final boolean isClearAggregationKeyOverride)
+	{
+		candidate.setM_Material_Tracking_ID(-1);
+		if (isClearAggregationKeyOverride)
+		{
+			candidate.setC_Invoice_Candidate_HeaderAggregation_Override(null);
+		}
+		aggregationBL.getUpdateProcessor().process(candidate);
+
+		repo.save(candidate);
+
+		materialTrackingBL.unlinkModelFromMaterialTrackings(candidate);
+	}
+
+	public void linkMaterialTrackings(@NonNull final List<I_C_Invoice_Candidate> candidate, @NonNull final MaterialTrackingId materialTrackingId)
+	{
+		final I_M_Material_Tracking materialTracking = materialTrackingDAO.getById(materialTrackingId);
+		if (materialTracking.isProcessed())
+		{
+			//target material tracking is processed, so we can't link to it
+			return;
+		}
+		candidate.forEach(c -> linkMaterialTracking(c, materialTracking));
+	}
+
+	public void linkMaterialTracking(@NonNull final I_C_Invoice_Candidate candidate, @NonNull final I_M_Material_Tracking materialTracking)
+	{
+		final MaterialTrackingId oldMaterialTrackingId = MaterialTrackingId.ofRepoIdOrNull(candidate.getM_Material_Tracking_ID());
+		if (oldMaterialTrackingId != null)
+		{
+			final I_M_Material_Tracking oldMaterialTracking = materialTrackingDAO.getById(oldMaterialTrackingId);
```

Comment:
bonus points: avoiding SQL N+1 would be nice....
but i understand that there is no so much benefit, since couple of lines below we have getFirstForMaterialTrackingId which is introducing anther SQL N+1 issue.